google-cloud-firestore/src/test/java/com/google/cloud/firestore/QueryTest.java

```diff
@@ -952,6 +953,76 @@ public void onCompleted() {
     semaphore.acquire();
   }
 
+  @Test
+  public void successfulReturnWithoutOnComplete() throws Exception {
+    doAnswer(queryResponseWithDone(true, DOCUMENT_NAME + "1", DOCUMENT_NAME + "2"))
+        .when(firestoreMock)
+        .streamRequest(
+            runQuery.capture(),
+            streamObserverCapture.capture(),
+            Matchers.<ServerStreamingCallable>any());
+
+    final Semaphore semaphore = new Semaphore(0);
+    final Iterator<String> iterator = Arrays.asList("doc1", "doc2").iterator();
+
+    query.stream(
+        new ApiStreamObserver<DocumentSnapshot>() {
+          @Override
+          public void onNext(DocumentSnapshot documentSnapshot) {
+            assertEquals(iterator.next(), documentSnapshot.getId());
+          }
+
+          @Override
+          public void onError(Throwable throwable) {
+            fail();
+          }
+
+          @Override
+          public void onCompleted() {
+            semaphore.release();
+          }
+        });
+
+    semaphore.acquire();
+  }
+
+  @Test
+  /**
+   * onComplete() will be called twice. The first time is when it detects RunQueryResponse.done set
+   * to true. The second time is when it receives half close
+   */
+  public void successfulReturnCallsOnCompleteTwice() throws Exception {
+    doAnswer(queryResponseWithDone(false, DOCUMENT_NAME + "1", DOCUMENT_NAME + "2"))
+        .when(firestoreMock)
+        .streamRequest(
+            runQuery.capture(),
+            streamObserverCapture.capture(),
+            Matchers.<ServerStreamingCallable>any());
+
+    final Semaphore semaphore = new Semaphore(0);
+    final Iterator<String> iterator = Arrays.asList("doc1", "doc2").iterator();
+
+    query.stream(
+        new ApiStreamObserver<DocumentSnapshot>() {
+          @Override
+          public void onNext(DocumentSnapshot documentSnapshot) {
+            assertEquals(iterator.next(), documentSnapshot.getId());
+          }
+
+          @Override
+          public void onError(Throwable throwable) {
+            fail();
+          }
+
+          @Override
+          public void onCompleted() {
+            semaphore.release();
+          }
+        });
+
+    semaphore.acquire();
```

Comment:
We should verify onComplete is really only called once. Here it is not clear we are checking this. if `semaphore.release()` is called twice the test will still pass.

But the API does not give us a good way to verify it is not called a second time. I cannot think of a way to do this without introducing some timer.

What we can do is probably in `onCompleted()`, check to see if a flag is set from previous call, and throw if so. Then set that flag so we know it has been called.

This is not guaranteed to catch the problem, but at least we tried. We should also add a comment to explain why if you choose to do so.